h2o-algos/src/main/java/hex/glrm/GLRM.java

```diff
@@ -1242,57 +1222,39 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
         // Categorical columns
         for (int j = 0; j < _ncats; j++) {
           if (Double.isNaN(a[j])) continue;   // Skip missing observations in row
-//          double[] xy = ArrayUtils.multVecArr(xnew, _yt.getCatBlock(j));
-//          _loss +=  _lossFunc[j].mloss(xy, (int) a[j]);
-
-          double[] txy = multVecArrFast(xnew, _yt, j);
-          _loss +=  _lossFunc[j].mloss(txy, (int) a[j]);
-
-          // check to see if xy and txy are the same, comment it out, do not delete yet.
-/*          double diff = 0.0;
-          for (int index = 0; index < txy.length; index++)
-            diff += xy[index]-txy[index];
-          assert diff < 1e-10;*/
+          double[] txy = multVecArrFast(xnew, prod, _yt, j);
+          _loss +=  _lossFunc[j].mloss(txy, (int) a[j], _yt._numLevels[j]);
         }
 
         // Numeric columns
         for (int j = _ncats; j < _ncolA; j++) {
           int js = j - _ncats;
           if (Double.isNaN(a[j])) continue;   // Skip missing observations in row
-          double xy = _yt.lmulNumCol(xnew, js);
-          _loss += _lossFunc[j].loss(xy, (a[j] - _normSub[js]) * _normMul[js]);
+          double txy = _yt.lmulNumCol(xnew, js);
+          _loss += _lossFunc[j].loss(txy, (a[j] - _normSub[js]) * _normMul[js]);
         }
         _loss *= cweight;
       }
     }
 
 
     /* same as ArrayUtils.multVecArr() but faster I hope. */
-    private double[] multVecArrFast(double[] xnew, Archetypes yt, int j) {
-      double[] xy = new double[yt._numLevels[j]];
+    private double[] multVecArrFast(double[] xnew, double[] xy, Archetypes yt, int j) {
+      Arrays.fill(xy, 0.0);
       if (yt._transposed) {
         for (int level = 0; level < yt._numLevels[j]; level++) {
           int cidx = yt.getCatCidx(j, level);
           double[] yArchetypes = yt._archetypes[cidx];
-          for (int k = 0; k < _ncolX; k++) {
-//            double archValue = 0.0;
-//            archValue = yt._archetypes[cidx][k];
-//            xy[level] += xnew[k] * archValue;
+          for (int k = 0; k < _ncolX; k++)
             xy[level] += xnew[k] * yArchetypes[k];
-          }
         }
       } else {
         for (int level = 0; level < yt._numLevels[j]; level++) {
           int cidx = yt.getCatCidx(j, level);
-          for (int k = 0; k < _ncolX; k++) {
-//            double archValue = 0.0;
-//            archValue = yt._archetypes[k][cidx];
-//            xy[level] += xnew[k] * archValue;
+          for (int k = 0; k < _ncolX; k++)
             xy[level] += xnew[k] * yt._archetypes[k][cidx];
-          }
         }
       }
-
       return xy;
```

Comment:
I would recommend not to return anything -- this way it's much clearer that the function writes directly into the provided array